### ISSUE TITLE
Is there an option to ignore background images during training?

### ISSUE THREAD

**Issue by IlkayW** [ultralytics/yolov5#11590]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hiho,

I'm currently trying to train a model based on some COCO classes (handbag, backpack, suitcase, etc.). 
However, my COCO folder contains the whole dataset and only the txt files of the selected classes.
This results into $\approx$ 37K labeled images and $\approx$ 80K background images.

Is there a way to ignore the unlabeled/background images or do I have to delete the / move them into a separate folder?
(The first way would be way more comfortable for me)

Cheers,
Ilkay

**Comment by glenn-jocher** [comment]
@IlkayW hi Ilkay,

Yes, you can ignore unlabeled/background images during training. You can create a custom dataset class that uses only images with labeled objects, or you can edit the train.txt file to include only the filenames of labeled images.

For example, you can use something like this to filter out images without labels:

```python
import os

with open('train.txt', 'r') as f:
    lines = f.readlines()

filtered_lines = [line for line in lines if os.path.isfile(line.strip().replace('images', 'labels').replace('.jpg', '.txt'))]

with open('train_filtered.txt', 'w') as f:
    f.writelines(filtered_lines)
```

This will create a new file called `train_filtered.txt` that only includes images with corresponding label files.

Hope this helps! Let me know if you have any other questions.

Best, 
Glenn

**Comment by IlkayW** [comment]
Thank you!
In data.yaml
train: path/to/train.txt
val: path/to/val.txt


But, I feel this option could be good anyways if the path to the train images is passed: data.yaml
train: path/to/train_images/
val: path/to/val_images/
and 
python train.py ... --ignore_background

**Comment by glenn-jocher** [comment]
@IlkayW hello,

Thank you for your suggestion! We appreciate your input in making YOLOv5 even better. 

Regarding your suggestion, having an `--ignore_background` option would be a useful feature, and we will consider it for implementation in the future. However, for now, you can achieve this by creating a custom dataset class that uses only labeled images or editing the `train.txt` file to include only the filenames of labeled images, as we discussed in a previous issue.

Please let us know if you have any further suggestions or questions. Thank you again for reaching out, and we hope you enjoy using YOLOv5!

Best, 
Glenn

**Comment by usapeshka** [comment]
@glenn-jocher thank you for your answer. 
Could you please clarify on 

> you can achieve this by creating a custom dataset class that uses only labeled images 

For example I have a dataset where I have 100 images and 50 of them are labeled, therefore, 50 are background(non labeled) images. Let's say that I currently have just 2 classes - [0] pizza and [1] hot_dog. How will the custom data class look like ?

Thank you in advance!!!

**Comment by glenn-jocher** [comment]
@usapeshka,

Great question and happy to help!

In your case, you can create a custom dataset class that filters out unlabeled images. Here is an example implementation:

```python
import os
from PIL import Image
import torch.utils.data as data


class CustomDataset(data.Dataset):

    def __init__(self, image_dir, label_dir, transform=None):
        self.transform = transform

        # get labeled image file paths
        label_files = os.listdir(label_dir)
        label_files = [f for f in label_files if f.endswith('.txt')]
        label_paths = [os.path.join(label_dir, f) for f in label_files]
        labeled_images = [os.path.splitext(f)[0] for f in label_files]

        # get selected classes
        classes = ['pizza', 'hot_dog']

        # get labeled images with selected classes
        self.image_paths = []
        for img_path in os.listdir(image_dir):
            if img_path.startswith('.') or img_path.endswith('_mask.jpg'):
                continue
            img_name = os.path.splitext(img_path)[0]
            if img_name not in labeled_images:
                continue
            label_path = os.path.join(label_dir, img_name + '.txt')
            with open(label_path, 'r') as f:
                labels = f.readlines()
            labels = [x.strip().split() for x in labels]
            labels = [x for x in labels if x[0] in classes]
            if labels:
                self.image_paths.append(os.path.join(image_dir, img_path))

    def __getitem__(self, index):
        image_path = self.image_paths[index]
        image = Image.open(image_path).convert('RGB')

        if self.transform is not None:
            image = self.transform(image)

        return image

    def __len__(self):
        return len(self.image_paths)
```

In this example, we filter out images without labels and then further filter the remaining images to include only those that contain the selected classes (`pizza` and `hot_dog`). Note that this example assumes that the label files are in YOLO format, where each line contains the class label and bounding box coordinates.

With this custom dataset class, you can pass it to the `DataLoader` in your training script, like so:

```python
train_dataset = CustomDataset(image_dir='path/to/train/images', label_dir='path/to/train/labels', transform=None
